std/algorithm/comparison.d

```diff
@@ -1565,3 +1569,367 @@ unittest
                 2, "two",
                 ));
 }
+
+/**
+Returns $(D true) if both ranges have the same length. This function is optimized
+to always take advantage of the $(D length) member of either range if it exists.
+
+If both ranges have a length member, this function is $(BIGOH 1). Otherwise,
+this function is $(BIGOH min(r1.length, r2.length)).
+
+Params:
+    r1 = a finite input range
+    r2 = a finite input range
+
+Returns:
+    $(D bool)
+*/
+bool isSameLength(Range1, Range2)(Range1 r1, Range2 r2)
+    if (isInputRange!Range1 &&
+        isInputRange!Range2 &&
+        !isInfinite!Range1 &&
+        !isInfinite!Range2)
+{
+    static if (hasLength!(Range1) && hasLength!(Range2))
+    {
+        if (r1.length != r2.length)
+        {
+            return false;
+        }
+    }
+    else static if (hasLength!(Range1) && !hasLength!(Range2))
+    {
+        size_t length;
+
+        while (!r2.empty)
+        {
+            r2.popFront;
+
+            if (++length > r1.length)
+            {
+                return false;
+            }
+        }
+
+        if (length < r1.length)
+        {
+            return false;
+        }
```

Comment:
`length < r1.length` can't be turned into a simple return statement, as `length > r1.length` is false as well.